README.md
# cnj-common-rest-jee

Provides common REST API classes to simplify implementation of JAX-RS based endpoints.

## Status
![Build status](https://drone.cloudtrain.aws.msgoat.eu/api/badges/msgoat/cnj-common-rest-jee/status.svg)

## Tracing inbound and outbound requests

Tracing of inbound and outbound requests is disabled by default. To enable tracing, simply set configuration key
`cloudtrain.common.rest.trace.enabled` to __true__.

All traces are logged using logger `cloudtrain.common.rest.trace`, which is expected to be set to logging level __INFO__.

## Release information

Latest version is: `1.1.0`.

A changelog can be found in [changelog.md](changelog.md).


changelog.md
# Changelog
All notable changes to cnj-common-rest-jee will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.1.0] - 2021-04-21
### Added
### Changed
- performed lift&shift to new AWS CloudTrain environment
- update to latest versions of runtime environments pending

## [1.1.0] - 2020-07-21
### Added
- added JwtPropagationClientRequestFilter for automatic propagation of JWTs to downstream services
### Changed
- reworked all ClientRequestFilters and ClientResponseFilters to support proper functionality in Payara

## [1.0.0] - 2020-07-01
### Added
- first revision
### Changed


src/main/java/group/msg/at/cloud/common/rest/internal/json/SimpleJsonBuilder.java
package group.msg.at.cloud.common.rest.internal.json;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.List;

public final class SimpleJsonBuilder {

    private static final String DOUBLE_QUOTE = "\"";
    private static final String VALUE_SEPARATOR = ":";
    private static final String ELEMENT_SEPARATOR = ", ";

    private final StringBuilder document = new StringBuilder();
    private final Deque<Integer> elements = new ArrayDeque<>();
    private int currentElementIndex = 0;

    public SimpleJsonBuilder add(String name, String value) {
        if (value != null && !value.isEmpty()) {
            if (currentElementIndex != 0) {
                document.append(ELEMENT_SEPARATOR);
            }
            currentElementIndex++;
            document.append(DOUBLE_QUOTE).append(name).append(DOUBLE_QUOTE).append(VALUE_SEPARATOR).append(DOUBLE_QUOTE).append(value).append(DOUBLE_QUOTE);
        }
        return this;
    }

    public SimpleJsonBuilder add(String name, int value) {
        if (currentElementIndex != 0) {
            document.append(ELEMENT_SEPARATOR);
        }
        currentElementIndex++;
        document.append(DOUBLE_QUOTE).append(name).append(DOUBLE_QUOTE).append(VALUE_SEPARATOR).append(value);
        return this;
    }

    public SimpleJsonBuilder add(String name, List<String> values) {
        if (values != null && !values.isEmpty()) {
            if (values.size() == 1) {
                add(name, values.get(0));
            } else {
                if (currentElementIndex != 0) {
                    document.append(ELEMENT_SEPARATOR);
                }
                currentElementIndex++;
                document.append(DOUBLE_QUOTE).append(name).append(DOUBLE_QUOTE).append(VALUE_SEPARATOR).append("[");
                int currentValueIndex = 0;
                for (String currentValue : values) {
                    if (currentValueIndex > 0) {
                        document.append(",");
                    }
                    document.append(DOUBLE_QUOTE).append(currentValue).append(DOUBLE_QUOTE);
                    currentValueIndex++;
                }
                document.append("]");
            }
        }
        return this;
    }

    public SimpleJsonBuilder startObject(String name) {
        if (currentElementIndex != 0) {
            document.append(ELEMENT_SEPARATOR);
        }
        elements.push(++currentElementIndex);
        currentElementIndex = 0;
        document.append(DOUBLE_QUOTE).append(name).append(DOUBLE_QUOTE).append(" : { ");
        return this;
    }

    public SimpleJsonBuilder stopObject(String name) {
        document.append(" }");
        currentElementIndex = elements.pop();
        return this;
    }

    public SimpleJsonBuilder startMap(String name) {
        if (currentElementIndex != 0) {
            document.append(ELEMENT_SEPARATOR);
        }
        elements.push(++currentElementIndex);
        currentElementIndex = 0;
        document.append(DOUBLE_QUOTE).append(name).append(DOUBLE_QUOTE).append(" : { ");
        return this;
    }

    public SimpleJsonBuilder stopMap(String name) {
        document.append(" }");
        currentElementIndex = elements.pop();
        return this;
    }

    public String build() {
        return document.toString();
    }
}


src/main/java/group/msg/at/cloud/common/web/jaxrs/RouterAwareUriBuilderFactory.java
package group.msg.at.cloud.common.web.jaxrs;

import javax.ws.rs.core.HttpHeaders;
import javax.ws.rs.core.MultivaluedMap;
import javax.ws.rs.core.UriBuilder;
import javax.ws.rs.core.UriInfo;
import java.net.URI;
import java.net.URISyntaxException;
import java.util.List;
import java.util.Optional;

/**
 * {@code Factory} for {@code UriBuilder}s which are aware of routers that forwarded the current request.
 * <p>
 * Use this class instead of the factory methods provided by {@link UriInfo} or {@link UriBuilder}, if your service
 * is running behind a router that uses path-based routing.
 * </p>
 *
 * @TODO: clarify if this factory can be transformed into a @RequestScoped bean with injectable UriInfo and HttpHeaders
 */
public class RouterAwareUriBuilderFactory {

    public static UriBuilder from(UriInfo uriInfo, HttpHeaders httpHeaders) {
        URI actualUri = null;
        try {
            actualUri = new URI(getScheme(uriInfo, httpHeaders), getAuthority(uriInfo, httpHeaders), getPath(uriInfo, httpHeaders), uriInfo.getRequestUri().getQuery(), uriInfo.getRequestUri().getFragment());
        } catch (URISyntaxException ex) {
            throw new IllegalStateException(String.format("failed to create a external URI for request [%s]", uriInfo.getRequestUri()), ex);
        }
        return UriBuilder.fromUri(actualUri);
    }

    private static String getScheme(UriInfo uriInfo, HttpHeaders headers) {
        Optional<String> forwardedPrefix = getHeaderByNameIgnoringCase(headers, "X-Forwarded-Proto");
        return forwardedPrefix.orElse(uriInfo.getRequestUri().getScheme());
    }

    private static String getAuthority(UriInfo uriInfo, HttpHeaders headers) {
        Optional<String> forwardedPrefix = getHeaderByNameIgnoringCase(headers, "X-Forwarded-Host");
        return forwardedPrefix.orElse(uriInfo.getRequestUri().getAuthority());
    }

    private static String getPath(UriInfo uriInfo, HttpHeaders headers) {
        Optional<String> forwardedPrefix = getHeaderByNameIgnoringCase(headers, "X-Forwarded-Prefix");
        return forwardedPrefix.orElse("") + uriInfo.getRequestUri().getPath();
    }

    private static Optional<String> getHeaderByNameIgnoringCase(HttpHeaders headers, String name) {
        String result = null;
        for (MultivaluedMap.Entry<String, List<String>> current : headers.getRequestHeaders().entrySet()) {
            if (current.getKey().equalsIgnoreCase(name)) {
                result = current.getValue() != null && !current.getValue().isEmpty() ? current.getValue().get(0) : null;
            }
        }
        return Optional.ofNullable(result);
    }
}


src/main/java/group/msg/at/cloud/common/rest/tracing/TracingClientRequestFilter.java
package group.msg.at.cloud.common.rest.tracing;

import group.msg.at.cloud.common.rest.internal.json.SimpleJsonBuilder;
import org.eclipse.microprofile.config.Config;
import org.eclipse.microprofile.config.ConfigProvider;
import org.eclipse.microprofile.config.inject.ConfigProperty;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.annotation.Priority;
import javax.inject.Inject;
import javax.ws.rs.Priorities;
import javax.ws.rs.client.ClientRequestContext;
import javax.ws.rs.client.ClientRequestFilter;
import javax.ws.rs.core.MultivaluedMap;
import javax.ws.rs.ext.Provider;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

/**
 * {@code JAX-RS ClientRequestFilter} which traces outbound requests to downstream services.
 * <p>
 * Since Payara (5.2020) does not handle CDI-Injections into ClientRequestFilters registered via {@code @RegisterProvider}
 * on MicroProfile REST clients very well, the actual MicroProfile configuration values must be looked up programmatically.
 * </p>
 * <p>
 * Quarkus (1.5) complains about CDI injection into JAX-RS providers but supports it nevertheless.
 * </p>
 * <p>
 * <strong>Attention:</strong> With Payara, this {@code ClientRequestFilter} has to be registered explicitly on
 * MicroProfile REST clients to be actually applied to REST client invocations. Quarkus picks all ClientRequestFilters
 * automatically.
 * </p>
 */
@Provider
@Priority(Priorities.USER)
public class TracingClientRequestFilter implements ClientRequestFilter {

    private static final Logger LOGGER = LoggerFactory.getLogger(Constants.TRACE_LOGGER_NAME);

    Boolean enabled;

    @Override
    public void filter(ClientRequestContext requestContext) throws IOException {
        if (shouldFilter(requestContext)) {
            SimpleJsonBuilder builder = new SimpleJsonBuilder();
            traceRequest(builder, requestContext);
            LOGGER.info("\"sentOutboundRequest\" : { {} }", builder.build());
        }
    }

    private boolean shouldFilter(ClientRequestContext requestContext) {
        return isEnabled() && LOGGER.isInfoEnabled();
    }

    private void traceRequest(SimpleJsonBuilder builder, ClientRequestContext requestContext) {
        builder.startObject("request");
        builder.add("uri", requestContext.getUri().toString());
        builder.add("method", requestContext.getMethod());
        builder.startMap("headers");
        MultivaluedMap<String, Object> headers = requestContext.getHeaders();
        headers.forEach((k, v) -> {
            List<String> values = new ArrayList<>();
            v.forEach(e -> values.add(e.toString()));
            builder.add(k, values);
        });
        builder.stopMap("headers");
        builder.stopObject("request");
    }

    private boolean isEnabled() {
        if (enabled == null) {
            ConfigProvider.getConfig().getOptionalValue(Constants.ENABLED_CONFIG_KEY, Boolean.class).ifPresentOrElse(v -> enabled = v, () -> enabled = Boolean.FALSE);
        }
        return enabled;
    }
}


src/main/java/group/msg/at/cloud/common/rest/jwt/JwtPropagatingClientRequestFilter.java
package group.msg.at.cloud.common.rest.jwt;

import org.eclipse.microprofile.jwt.JsonWebToken;

import javax.annotation.Priority;
import javax.enterprise.inject.spi.CDI;
import javax.ws.rs.Priorities;
import javax.ws.rs.client.ClientRequestContext;
import javax.ws.rs.client.ClientRequestFilter;
import javax.ws.rs.ext.Provider;
import java.io.IOException;

/**
 * {@code JAX-RS ClientRequestFilter} that propagates an existing {@link JsonWebToken} to a downstream service.
 * <p>
 * Since Payara (5.2020) does not handle CDI-Injections into ClientRequestFilters registered via {@code @RegisterProvider}
 * on MicroProfile REST clients very well, the actual {@code JsonWebToken} object must be looked up programmatically.
 * </p>
 * <p>
 * Quarkus (1.5) complains about CDI injection into JAX-RS providers but supports it nevertheless.
 * </p>
 * <p>
 * <strong>Attention:</strong> With Payara, this {@code ClientRequestFilter} has to be registered explicitly on
 * MicroProfile REST clients to be actually applied to REST client invocations. Quarkus picks all ClientRequestFilters
 * automatically.
 * </p>
 */
@Provider
@Priority(Priorities.HEADER_DECORATOR)
public class JwtPropagatingClientRequestFilter implements ClientRequestFilter {

    @Override
    public void filter(ClientRequestContext requestContext) throws IOException {
        JsonWebToken jwt = CDI.current().select(JsonWebToken.class).get();
        if (jwt != null) {
            propagateAuthorizationHeader(requestContext, jwt);
        }
    }

    private void propagateAuthorizationHeader(ClientRequestContext requestContext, JsonWebToken jwt) {
        requestContext.getHeaders().putSingle("Authorization", "Bearer " + jwt.getRawToken());
    }
}


src/main/java/group/msg/at/cloud/common/rest/tracing/TracingContainerRequestFilter.java
package group.msg.at.cloud.common.rest.tracing;

import group.msg.at.cloud.common.rest.internal.json.SimpleJsonBuilder;
import org.eclipse.microprofile.config.inject.ConfigProperty;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.inject.Inject;
import javax.ws.rs.container.ContainerRequestContext;
import javax.ws.rs.container.ContainerRequestFilter;
import javax.ws.rs.core.MultivaluedMap;
import javax.ws.rs.ext.Provider;
import java.io.IOException;

/**
 * {@code JAX-RS ContainerRequestFilter} which traces inbound requests from upstream services or frontends.
 * <p>
 * In contrast to ClientRequestFilters both CDI injection and automatic detection works well for ContainerRequestFilters
 * on all MicroProfile capable application servers.
 * </p>
 */
@Provider
public class TracingContainerRequestFilter implements ContainerRequestFilter {

    private static final Logger LOGGER = LoggerFactory.getLogger(Constants.TRACE_LOGGER_NAME);

    @Inject
    @ConfigProperty(name = Constants.ENABLED_CONFIG_KEY, defaultValue = Constants.ENABLED_DEFAULT_VALUE)
    boolean enabled;

    @Override
    public void filter(ContainerRequestContext requestContext) throws IOException {
        if (shouldFilter(requestContext)) {
            SimpleJsonBuilder builder = new SimpleJsonBuilder();
            traceRequest(builder, requestContext);
            LOGGER.info("\"receivedInboundRequest\" : { {} }", builder.build());
        }
    }

    private boolean shouldFilter(ContainerRequestContext requestContext) {
        return enabled && LOGGER.isInfoEnabled() && !requestContext.getUriInfo().getPath().contains("probes");
    }

    private void traceRequest(SimpleJsonBuilder builder, ContainerRequestContext request) {
        builder.startObject("request");
        builder.add("uri", request.getUriInfo().getRequestUri().toString());
        builder.add("method", request.getMethod());
        builder.startMap("headers");
        MultivaluedMap<String, String> headers = request.getHeaders();
        headers.forEach((k, v) -> builder.add(k, v));
        builder.stopMap("headers");
        builder.stopObject("request");
    }
}


src/main/java/group/msg/at/cloud/common/rest/tracing/TracingContainerResponseFilter.java
package group.msg.at.cloud.common.rest.tracing;

import group.msg.at.cloud.common.rest.internal.json.SimpleJsonBuilder;
import org.eclipse.microprofile.config.inject.ConfigProperty;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.inject.Inject;
import javax.ws.rs.container.ContainerRequestContext;
import javax.ws.rs.container.ContainerResponseContext;
import javax.ws.rs.container.ContainerResponseFilter;
import javax.ws.rs.ext.Provider;
import java.io.IOException;

/**
 * {@code JAX-RS ContainerResponseFilter} which traces outbound responses to upstream services or frontends.
 * <p>
 * In contrast to ClientRequestFilters both CDI injection and automatic detection works well for ContainerRequestFilters
 * on all MicroProfile capable application servers.
 * </p>
 */
@Provider
public class TracingContainerResponseFilter implements ContainerResponseFilter {

    private static final Logger LOGGER = LoggerFactory.getLogger(Constants.TRACE_LOGGER_NAME);

    @Inject
    @ConfigProperty(name = Constants.ENABLED_CONFIG_KEY, defaultValue = Constants.ENABLED_DEFAULT_VALUE)
    boolean enabled;

    @Override
    public void filter(ContainerRequestContext requestContext, ContainerResponseContext responseContext) throws IOException {
        if (shouldFilter(requestContext, responseContext)) {
            SimpleJsonBuilder builder = new SimpleJsonBuilder();
            traceRequest(builder, requestContext);
            traceResponse(builder, responseContext);
            LOGGER.info("\"sentOutboundResponse\" : { {} }", builder.build());
        }
    }

    private boolean shouldFilter(ContainerRequestContext requestContext, ContainerResponseContext responseContext) {
        return enabled && LOGGER.isInfoEnabled() && !requestContext.getUriInfo().getPath().contains("probes");
    }

    private void traceRequest(SimpleJsonBuilder builder, ContainerRequestContext requestContext) {
        builder.startObject("request");
        builder.add("uri", requestContext.getUriInfo().getRequestUri().toString());
        builder.add("method", requestContext.getMethod());
        builder.stopObject("request");
    }

    private void traceResponse(SimpleJsonBuilder builder, ContainerResponseContext responseContext) {
        builder.startObject("response");
        builder.add("status", responseContext.getStatus());
        builder.stopObject("response");
    }

}


src/main/java/group/msg/at/cloud/common/rest/tracing/TracingClientResponseFilter.java
package group.msg.at.cloud.common.rest.tracing;

import group.msg.at.cloud.common.rest.internal.json.SimpleJsonBuilder;
import org.eclipse.microprofile.config.ConfigProvider;
import org.eclipse.microprofile.config.inject.ConfigProperty;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.annotation.Priority;
import javax.inject.Inject;
import javax.ws.rs.Priorities;
import javax.ws.rs.client.ClientRequestContext;
import javax.ws.rs.client.ClientResponseContext;
import javax.ws.rs.client.ClientResponseFilter;
import javax.ws.rs.ext.Provider;
import java.io.IOException;

/**
 * {@code JAX-RS ClientResponseFilter} which traces inbound responses from downstream services.
 * <p>
 * Since Payara (5.2020) does not handle CDI-Injections into ClientResponseFilters registered via {@code @RegisterProvider}
 * on MicroProfile REST clients very well, the actual MicroProfile configuration values must be looked up programmatically.
 * </p>
 * <p>
 * Quarkus (1.5) complains about CDI injection into JAX-RS providers but supports it nevertheless.
 * </p>
 * <p>
 * <strong>Attention:</strong> With Payara, this {@code ClientResponseFilter} has to be registered explicitly on
 * MicroProfile REST clients to be actually applied to REST client invocations. Quarkus picks all ClientResponseFilters
 * automatically.
 * </p>
 */
@Provider
@Priority(Priorities.USER)
public class TracingClientResponseFilter implements ClientResponseFilter {

    private static final Logger LOGGER = LoggerFactory.getLogger(Constants.TRACE_LOGGER_NAME);

    Boolean enabled;

    private boolean shouldFilter(ClientRequestContext requestContext, ClientResponseContext responseContext) {
        return isEnabled() && LOGGER.isInfoEnabled();
    }

    @Override
    public void filter(ClientRequestContext requestContext, ClientResponseContext responseContext) throws IOException {
        if (shouldFilter(requestContext, responseContext)) {
            SimpleJsonBuilder builder = new SimpleJsonBuilder();
            traceRequest(builder, requestContext);
            traceResponse(builder, responseContext);
            LOGGER.info("\"receivedInboundResponse\" : { {} }", builder.build());
        }
    }

    private void traceRequest(SimpleJsonBuilder builder, ClientRequestContext requestContext) {
        builder.startObject("request");
        builder.add("uri", requestContext.getUri().toString());
        builder.add("method", requestContext.getMethod());
        builder.stopObject("request");
    }

    private void traceResponse(SimpleJsonBuilder builder, ClientResponseContext responseContext) {
        builder.startObject("response");
        builder.add("status", responseContext.getStatus());
        builder.stopObject("response");
    }

    private boolean isEnabled() {
        if (enabled == null) {
            ConfigProvider.getConfig().getOptionalValue(Constants.ENABLED_CONFIG_KEY, Boolean.class).ifPresentOrElse(v -> enabled = v, () -> enabled = Boolean.FALSE);
        }
        return enabled;
    }

}
